### ISSUE TITLE
Allow making any widget clickable

### ISSUE THREAD

**Issue by epruesse** [home-assistant/appdaemon#282]
It would be nice if it was possible to assign an onclick action to any widget. In particular, the navigation action would be useful. 

My use case would be having overview type widgets (clock, weather, camera, device_tracker, ...) switch to another dash with more detail. E.g. have a small camera widget and show a fullscreen dash for that cam if it is clicked.

**Comment by kabturek** [comment]
what other actions apart from navigate are you guys thinking about ?

**Comment by epruesse** [comment]
- rotate through alternate widgets in the same slot (timed, on click)
   - current weather replaced by forecast
   - rotate through cameras
   - current time replaced by relative's time zone
- initiate detail view (full screen, navigation or popup type reaction)
  - all the weather data on click on the weather widget
  - bigger security camera view, potentially showing this like a history
  - map on click of a device tracker widget
- map alternate input actions
  - multi-click on light to set a specific scene (hue remote does this)
  - long click to dim light
  - swipe to dim (I think there are JS apis to get swipes, not sure though)
 
If I understand the architecture of appdaemon right, most of this would be accomplishable by having a way to customize translation of JS events into HA/Appdaemon events.

**Comment by epruesse** [comment]
(Plus perhaps having a widget that has multiple sub-widgets it can show)

**Comment by kabturek** [comment]
initiate detail view -> navigate does exactly that (with timeout etc)
map alternate input actions  -> custom/or extended light widget
multiple sub-widgets -> we have "sub-dashboards" to simplify this http://appdaemon.readthedocs.io/en/latest/DASHBOARD_CREATION.html#advanced-dashboard-definition

**Comment by epruesse** [comment]
> initiate detail view -> navigate does exactly that (with timeout etc)
Yes, but how would I make say a `weather` widget call navigate? I can do this with two widgets:

```
weather_widget:
    widget_type: weather
weather_dash_widget
    widget_type: navigate
    title: Detailed Weather
    dashboard: weather_detail
```

What I'd like to do is use save the space for the extra button. It's natural to click on the weather widget to get more info on it, more so than looking for another button elsewhere on the dash.

> map alternate input actions -> custom/or extended light widget
I suppose here I was asking for a different feature. Some more actions besides `post_service_active` and `post_service_inactive`. Agreed, that's a different issue. 

> multiple sub-widgets -> we have "sub-dashboards" to simplify this 
Yes, technically that would work. It gets pretty redundant and cluttered quickly though.  Even using "sub-dashboards", I'll need an extra `.yaml` for the line(s) showing the alternate widget and another `.dash` for the assembled dash. Every widget with an alternate view needs one extra "sub-dashboard" for every sub-dash it's included on and one extra `.dash` for every dash it's on.

**Comment by kabturek** [comment]
> Yes, but how would I make say a weather widget call navigate? 
sorry i wasn't clear - this use case would be contained in the initial idea to add click/navigate to (all? most?) widget

I think the other two requests are a good idea but i would treat them as other features.

i like the initial navigate idea and might look into this if no one takes this.

**Comment by dcolley** [comment]
how about an "action | on_click | on_tab" declaration

```
my_action_declaration:
    type: [web | http]
    url: http[s]://my.security.cam.ip/login

my_camera_widget:
    click_action: [!]action my_action_declaration

**Comment by ReneTode** [comment]
the plan is to have predefined javascripts that can be assigned to possible widget actions.
so you can assign longpres(double click)s to navigate and short press(single click) to change a light switch. or swipe to a color change.

the javascripts will have the same as custom widgets. they can be designed by everyone, but there will be a lot of predefined actions.

the trouble is that we only start working on that after the release from AD 4 (which is way overdue through circumstances)

**Comment by DaveDixon** [comment]
Upvote. Navigation is what I need - I have two custom widgets, navsensor and navcamera. I use navsensor for navigation to/between alarm, lighting, and climate dashboards. I use navcamera to cycle through a list of cameras. I can't think of other use cases, since navigating to another dash always does the trick for me.
